crates/ntx-builder/src/transaction.rs

```diff
@@ -233,18 +246,38 @@ impl NtxContext {
 // ================================================================================================
 
 /// A [`DataStore`] implementation which provides transaction inputs for a single account and
-/// reference block.
+/// reference block with LRU caching for note scripts.
+///
+/// This implementation includes an LRU (Least Recently Used) cache for note scripts to improve
+/// performance by avoiding repeated RPC calls for the same script roots. The cache automatically
+/// manages memory usage by evicting least recently used entries when the cache reaches capacity.
 ///
 /// This is sufficient for executing a network transaction.
 struct NtxDataStore {
     account: Account,
     reference_header: BlockHeader,
     chain_mmr: PartialBlockchain,
     mast_store: TransactionMastStore,
+    /// RPC client for retrieving note scripts from the RPC server.
+    rpc_client: RpcClient,
+    /// LRU cache for storing retrieved note scripts to avoid repeated RPC calls.
+    script_cache: Arc<Mutex<LruCache<Word, miden_objects::note::NoteScript>>>,
 }
 
 impl NtxDataStore {
-    fn new(account: Account, reference_header: BlockHeader, chain_mmr: PartialBlockchain) -> Self {
+    /// Default cache size for note scripts.
+    ///
+    /// Each cached script contains the deserialized `NoteScript` object, so the actual memory usage
+    /// depends on the complexity of the scripts being cached.
+    const DEFAULT_SCRIPT_CACHE_SIZE: usize = 1000;
```

Comment:
This is fine for now, but we should make this a config parameter in the future. Let's create an issue for this.